### ISSUE TITLE
Private Module Function Not Found When Running All Test Scripts and Piping Multiple Objects

### ISSUE THREAD

**Issue by jsmythsci** [pester/Pester#2047]
## General summary of the issue
I am working on a module that wraps the API of one of our partners. As such, most functions just pass values to Invoke-RestMethod and return the response.

I am experiencing a strange issue where the private function that actually calls Invoke-RestMethod is returning a CommandNotFoundException but only when attempting to pass multiple objects in via the pipeline and only when I try to run the entire module's test suite.

## Describe your environment
Pester version     : 5.2.2 C:\Program Files\WindowsPowerShell\Modules\Pester\5.2.2\Pester.psm1
PowerShell version : 5.1.18362.1474
OS version         : Microsoft Windows NT 10.0.18363.0

Note that I was originally on v5.0.2 but updated before submitting this Issue. v5.0.2 has the same issue with the test failing but I can no longer confirm if the test passes when invoked directly on v5.0.2. I did not discover this behaviour until long after I had updated.

## Steps to reproduce
I am not really sure what the minimal steps are to reproduce this issue because it only happens when running the entire test suite. Running just the one test script does not cause the issue to trigger.

I will provide the information I have and maybe collectively we can whittle it down to isolate the issue.

Here is an example of a working test:
```powershell
    Context 'Pipelining a single object' {
        BeforeAll {
            Mock -CommandName $internalApiFunction { ($Body | ConvertFrom-Json) } -ModuleName $moduleName

            [PSCustomObject]$testInput = [PSCustomObject]@{
                Type = 'A'
                Name = 'foo.doma.in'
            }

            [Diagnostics.CodeAnalysis.SuppressMessageAttribute(
                'PSUserDeclaredVarsMoreThanAssignments',
                '',
                Justification = 'Variable is used in a different scope'
            )]
            $result = $testInput | Add-RecordViaApi
        }

        It 'Passes the correct "<DataType>" to the API' -TestCases @(
            @{DataType = 'Type'}
            @{DataType = 'Name'}
        ) {
            $expected = $testInput.$DataType

            $result.records[0].$DataType | Should -BeExactly $expected
        }
    }
```

Immediately below that Context block (in the same test script) is this one:
```powershell
    Context 'Pipelining multiple objects' {
        BeforeAll {
            Mock -CommandName $internalApiFunction { ($Body | ConvertFrom-Json) } -ModuleName $moduleName

            [PSCustomObject[]]$testInput = @(
                [PSCustomObject]@{ Type = 'A'; Name = 'foo.doma.in' }
                [PSCustomObject]@{ Type = 'A'; Name = 'bar.doma.in' }
            )

            [Diagnostics.CodeAnalysis.SuppressMessageAttribute(
                'PSUserDeclaredVarsMoreThanAssignments',
                '',
                Justification = 'Variable is used in a different scope'
            )]
            $result = $testInput | Add-RecordViaApi
        }

        It 'Invokes the API only once' {
            Should -Invoke -CommandName $internalApiFunction -Times 1 -Exactly
        }
    }
```

As far as I can tell, the only difference between the 2 tests is the number of objects being passed into the pipeline. The first test always passes but the latter fails when run as part of the full test suite:

```
[-] Add-RecordViaApi.Pipelining multiple objects.Invokes the API only once 450ms (447ms|3ms)
 
 CommandNotFoundException: Could not find Command Invoke-ApiAsync
```

Function `Invoke-ApiAsync` does exist; it is the function being Mocked by the `Mock -CommandName $internalApiFunction ...` lines in the snippets above 

This is how I usually invoke Pester during my workflow:
```
powershell.exe -NonInteractive -Command 'Invoke-Pester; gci -Recurse -File | Invoke-ScriptAnalyzer'; Get-Date
```

When I invoke Pester as above, the multi-object pipeline test fails. When I invoke just that one test script, the test passes:
```
powershell.exe -NonInteractive -Noprofile -Command 'Invoke-Pester -Path "./Tests/Add-RecordViaApi.Tests.ps1"'; Get-Date
```

## Expected Behavior
Test should pass regardless of how Pester is invoked.

## Current Behavior
When invoked directly, the test passes; when invoked with all test scripts (`Invoke-Pester` with no arguments), the test fails.

**Comment by fflaten** [comment]
I'm unable to reproduce with the current samples. Are you able to provide a working sample or at least the Diagnostics-output (`Invoke-Pester .....  -Output Diagnostics`)? Immediate thought would be that one of your other testfiles are calling `Remove-Module` to unload the module containing `Invoke-ApiSync`.

**Comment by jsmythsci** [comment]
Thank you for the feedback, @fflaten.

I ran some more tests and I have been able to determine that it is this line in the test that is causing the failure:
```powershell
Should -Invoke -CommandName $internalApiFunction -Times 1 -Exactly
```

In this context `$internalApiFunction` is set to the name of the private module function that the public function under test depends on. If I replace this assertion line with something generic (e.g. "$true"), then the test passes without issue.

This tells us that the issue is that the `CommandNotFoundException` is being thrown by the `Should` invocation. This seems reasonable because the command is private (not exported from the module) and `Should` is being run outside the module's context. I rewrote the assertion as follows and now it is working:

```powershell
        It 'Invokes the API only once' {
            [hashtable]$shouldParams = @{
                CommandName = $internalApiFunction
                Times = 1
                Exactly = $true
                ModuleName = $moduleName
                Scope = 'Context'
            }
            Should -Invoke @shouldParams
        }
```

Lastly, I also discovered that the test behaviour is consistent whether invoked as part of a full suite or as a standalone file; the reason I thought it passed when given a single file is because I was running the wrong test script (`Get-Foo` != `Add-Foo`).

With all of this, I think I have a good understanding of why the test was failing: I am still new to a lot of this and my design was flawed. After fixing the design, the test, and Pester in general, seems to be working as intended.
